cpp/examples/sbml_to_memilio.cpp

```diff
@@ -0,0 +1,921 @@
+            double value     = param.getValue();
+            std::string name = param.getId();
+            parameters << "template <typename FP = ScalarType>" << std::endl;
+            parameters << "struct " << name << " {\n    using Type = double;\n  static Type get_default()\n{"
+                       << std::endl;
+            parameters << "return " << value << ";\n}" << std::endl;
+            parameters << "static std::string name()\n{\n   return \"" << name << "\";\n}\n};\n" << std::endl;
+            parameterset_initializer = parameterset_initializer + name + "<FP>, ";
+        }
+    }
+    parameterset_initializer.pop_back();
+    parameterset_initializer.pop_back();
+    parameters << "template <typename FP = ScalarType>" << std::endl;
+    parameters << parameterset_initializer << ">;\n " << std::endl;
+    parameters << "template <typename FP = ScalarType>\nclass Parameters : public ParametersBase<FP>\n{\npublic:\n    "
+                  "Parameters()\n        : ParametersBase<FP>()\n    {\n    }\n\npublic:\n    template <class "
+                  "IOContext>\n    static IOResult<Parameters> deserialize(IOContext& io)\n    {\n        "
+                  "BOOST_OUTCOME_TRY(auto&& base, ParametersBase<FP>::deserialize(io));\n        return "
+                  "success(Parameters(std::move(base)));\n    }\n\n};\n}\n}\n#endif\n"
+               << std::endl;
+
+    parameters.close();
+
+    return true;
+}
+
+/*
+    @brief Create the model.cpp file
+
+    @param filename The name of the file to be processed
+
+    Creates the generic model.cpp file. The file location is generated using the filename.
+*/
+bool create_model_cpp(const char* filename)
+{
+
+    std::string lowercase_name = boost::to_lower_copy<std::string>(get_filename(filename));
+
+    std::ofstream model_cpp;
+    model_cpp.open(lowercase_name + "/model.cpp", std::ios::out);
+
+    model_cpp << "#include \"" << lowercase_name << "/model.h\"\n\nnamespace mio\n{" << std::endl;
+    model_cpp << "namespace " << lowercase_name << "\n{\n\n}\n}" << std::endl;
+    model_cpp.close();
+
+    return true;
+}
+
+/*
+    @brief Create the model.h file
+
+    @param model The model to be processed
+    @param filename The name of the file to be processed
+
+    Creates the model.h file based on the filename. First some generic input is written to the file. Then the `get_derivatives`-function is generated. It
+    - creates an index variable for every species
+    - creates a lambda function for every function definition in the model to have it at hand
+    - goes through every reaction in the model and generates the corresponding code. This is done in local scopes to make sure local parameters don't cause errors.
+    - checks every rule whether it is a RateRule and generates code for it if it is.
+    
+    In the end it appends a generic serialization function.
+*/
+bool create_model_h(Model* model, const char* filename)
```

Comment:
There is a lot of nested logic in this function, can you maybe move some of this into helper functions? Can some of it be reused? Because as is, this function is really hard to read and understand.

Optimally, I'd like to see no more than two indents, and have them closed as soon as possible. But do not take that as a requirement, only an aspiration.